lib/srv/reexec.go

```diff
@@ -418,6 +422,37 @@ func RunCommand() (errw io.Writer, code int, err error) {
 	return io.Discard, exitCode(err), trace.Wrap(err)
 }
 
+func waitForShell(termiantefd *os.File, cmd *exec.Cmd) error {
+	terminateChan := make(chan error)
+
+	go func() {
+		buf := make([]byte, 1)
+		// Wait for the terminate file descriptor to be closed.
+		_, err := termiantefd.Read(buf)
+		if err == io.EOF {
+			// Kill the shell process
+			err = cmd.Process.Kill()
+		} else {
+			err = trace.Errorf("failed to read from terminate file: %w", err)
+		}
+		terminateChan <- err
+	}()
+
+	go func() {
+		// Wait for the command to exit. It doesn't make sense to print an error
+		// message here because the shell has successfully started. If an error
+		// occurred during shell execution or the shell exits with an error (like
+		// running exit 2), the shell will print an error if appropriate and return
+		// an exit code.
+		err := cmd.Wait()
+
+		terminateChan <- err
+	}()
+
+	err := <-terminateChan
```

Comment:
You're doing 2 writes to `terminateChan` and only 1 read. Doesn't that mean that one of these goroutines will leak, and that you may miss an error depending on timing?

Reply:
> Doesn't that mean that one of these goroutines will leak

Yes, but after one goroutine finishes, the process ends, so leaking anything at this point is not a huge deal.

> and that you may miss an error depending on timing

This is more complicated. The ideal scenario happens when the shell process end, and we don't need to send kill signals to it. The scenario that this PR tries to fix happens when the SSH session sends on the client side, and someone needs to kill the "abandoned shell". If we try to kill the shell after the shell process ends, we don't care about any returned errors. I also need to check what is logged in our audit log when that happens (and TBH, I'm not sure what the correct value should be). If the shell process ends without an error, then we don't want to override that with the potential error returned from the other goroutine. I'll add a comment to explain that, as this is probably "not self-explanatory" just by looking at the code. I'm also not saying that the current version is 100% working.

Reply:
Yeah, your explanation here makes sense, it just isn't obvious in the code.